File: frontend-v2/src/lib/api/citationMap.ts
import { api } from './client';

export interface MapNodePosition {
  x: number;
  y: number;
}

export interface MapNode {
  key: string;
  s2_id: string | null;
  library_paper_id: number | null;
  title: string;
  authors: string;
  year: number | null;
  citation_count: number | null;
  is_focal: boolean;
  is_library: boolean;
  shared: boolean;
  doi: string | null;
  url: string | null;
  position: MapNodePosition | null;
}

export interface MapEdge {
  source: string;
  target: string;
  /** "reference" = focal paper cites target (what it built on). */
  type: 'reference';
}

export interface UnresolvedPaper {
  library_paper_id: number;
  title: string;
}

export interface CitationMapResponse {
  nodes: MapNode[];
  edges: MapEdge[];
  focal_paper_ids: number[];
  unresolved: UnresolvedPaper[];
}

export interface PositionUpdate {
  node_key: string;
  x: number;
  y: number;
}

export interface CitedByPaper {
  s2_id: string | null;
  title: string | null;
  authors: string[];
  year: number | null;
  citation_count: number | null;
  doi: string | null;
  url: string | null;
}

export interface CitedByResponse {
  paper_id: number;
  resolved: boolean;
  citations: CitedByPaper[];
  offset: number;
  limit: number;
  has_more: boolean;
}

export const citationMapApi = {
  get: (): Promise<CitationMapResponse> =>
    api.get<CitationMapResponse>('/citation-map'),

  addFocal: (paperId: number): Promise<CitationMapResponse> =>
    api.post<CitationMapResponse>('/citation-map/focal', { paper_id: paperId }),

  removeFocal: (paperId: number): Promise<CitationMapResponse> =>
    api.delete<CitationMapResponse>(`/citation-map/focal/${paperId}`),

  savePositions: (positions: PositionUpdate[]): Promise<void> =>
    api.post<void>('/citation-map/positions', { positions }),

  clear: (): Promise<void> => api.delete<void>('/citation-map'),

  citedBy: (paperId: number, offset = 0, limit = 25): Promise<CitedByResponse> =>
    api.get<CitedByResponse>(`/citation-map/cited-by/${paperId}`, {
      params: { offset, limit },
    }),
};

File: backend/app/schemas/citation_map.py
from pydantic import BaseModel


class NodePosition(BaseModel):
  x: float
  y: float


class MapNode(BaseModel):
  key: str
  s2_id: str | None = None
  library_paper_id: int | None = None
  title: str
  authors: str = ""
  year: int | None = None
  citation_count: int | None = None
  is_focal: bool = False
  is_library: bool = False
  shared: bool = False
  doi: str | None = None
  url: str | None = None
  position: NodePosition | None = None


class MapEdge(BaseModel):
  source: str
  target: str
  type: str  # "reference" — focal paper cites target


class UnresolvedPaper(BaseModel):
  library_paper_id: int
  title: str


class CitationMapResponse(BaseModel):
  nodes: list[MapNode]
  edges: list[MapEdge]
  focal_paper_ids: list[int]
  unresolved: list[UnresolvedPaper]


class FocalCreate(BaseModel):
  paper_id: int


class PositionUpdate(BaseModel):
  node_key: str
  x: float
  y: float


class BulkPositions(BaseModel):
  positions: list[PositionUpdate]


class CitedByPaper(BaseModel):
  s2_id: str | None = None
  title: str | None = None
  authors: list[str] = []
  year: int | None = None
  citation_count: int | None = None
  doi: str | None = None
  url: str | None = None


class CitedByResponse(BaseModel):
  paper_id: int
  resolved: bool
  citations: list[CitedByPaper]
  offset: int = 0
  limit: int = 25
  has_more: bool = False
